parser_library/src/context/using.cpp

```diff
@@ -0,0 +1,551 @@
+    int32_t min_disp,
+    int32_t max_disp,
+    bool ignore_length) const -> context_evaluate_result
+{
+    struct result_candidate
+    {
+        const entry* result_entry = nullptr;
+        ptrdiff_t result_reg = 0;
+        long long min_dist_abs = std::numeric_limits<long long>::max();
+        long long min_dist = 0;
+        int min_dist_reg = -1;
+    };
+    result_candidate positive;
+    result_candidate negative;
+
+    if (owner == nullptr && min_disp <= offset && offset <= max_disp)
+    {
+        // implicit 0 mapping
+        static constexpr entry zero_entry {
+            nullptr,
+            nullptr,
+            0,
+            0x1000,
+            []() {
+                register_set_t zero_reg = invalid_register_set;
+                zero_reg[0] = 0;
+                return zero_reg;
+            }(),
+            0,
+        };
+        (offset >= 0 ? positive : negative) = result_candidate { &zero_entry, 0, std::abs(offset), offset, 0 };
+    }
+
+    for (const auto& s : m_state)
+    {
+        if (label != s.label || owner != s.owner)
+            continue;
+        auto next_dist = (offset - s.offset) + s.reg_offset;
+        for (const auto& reg : s.regs)
+        {
+            const auto dist = next_dist;
+            next_dist -= 0x1000;
+
+            if (reg == invalid_register)
+                continue;
+
+            result_candidate& target = dist >= 0 ? positive : negative;
+            auto abs_dist = std::abs(dist);
+            if (abs_dist < target.min_dist_abs || abs_dist == target.min_dist_abs && reg > target.min_dist_reg)
+            {
+                target.min_dist_abs = abs_dist;
+                target.min_dist_reg = reg;
+                target.result_reg = &reg - s.regs.data();
+                target.result_entry = &s;
+                target.min_dist = dist;
+            }
+        }
+    }
+
+    const auto& r = [min_disp, max_disp](const auto& p, const auto& n) {
+        if (p.result_entry && p.min_dist >= min_disp && p.min_dist <= max_disp)
+            return p;
+        if (n.result_entry && n.min_dist >= min_disp && n.min_dist <= max_disp)
+            return n;
+
+        if (p.min_dist_abs <= n.min_dist_abs)
+            return p;
+        else
+            return n;
+    }(positive, negative);
+
+    if (!r.result_entry)
+        return context_evaluate_result { invalid_register_set, 0, 0 };
+
+    context_evaluate_result result {
+        invalid_register_set,
+        clamp<offset_t>(r.min_dist),
+        clamp<offset_t>(ignore_length ? 0 : r.result_entry->length - (offset - r.result_entry->offset)),
+    };
+
+    if (result.reg_offset >= min_disp && result.reg_offset <= max_disp)
+        std::copy(r.result_entry->regs.begin() + r.result_reg, r.result_entry->regs.end(), result.mapping_regs.begin());
+    else if (result.reg_offset < min_disp)
+        result.reg_offset -= min_disp;
+    else /* if (result.reg_offset > max_disp) */
+        result.reg_offset -= max_disp;
```

Comment:
I'm not sure what these two assignments should represent so my question might be irrelevant.

Nevertheless, it seems that when the `result.reg_offset < min_disp` the `min_disp` is subtracted from `result.reg_offset`. Shouldn't it be the other way around? Isn't the end goal to have `result.reg_offset` in the range `<min_disp, max_disp>`?

Reply:
I've been considering adding a union with something like out_of_range_by... because that's what it means when the mapping register set consists of invalid registers.

Reply:
I like this idea! Maybe it can be a feature for one of the next PRs